src/lib/support/BufferWriter.h

```diff
@@ -0,0 +1,194 @@
+    }
+
+    /// Returns whether the input fit in the buffer, outputs what was actually written
+    bool Fit(size_t & actuallyWritten) const
+    {
+        actuallyWritten = mSize >= mNeeded ? mNeeded : mSize;
+        return mSize >= mNeeded;
+    }
+
+    /// Size of the output buffer
+    size_t Size() const { return mSize; }
+
+    uint8_t * Buffer() { return mBuf; }
+    const uint8_t * Buffer() const { return mBuf; }
+
+protected:
+    friend Derived;
+
+    uint8_t * mBuf;
+    size_t mSize;
+    size_t mNeeded;
+
+    BufferWriterBase(uint8_t * buf, size_t len) : mBuf(buf), mSize(len), mNeeded(0) {}
+    BufferWriterBase(uint8_t * buf, size_t len, size_t needed) : mBuf(buf), mSize(len), mNeeded(needed) {}
+    BufferWriterBase(const BufferWriterBase & other) = default;
+    BufferWriterBase & operator=(const BufferWriterBase & other) = default;
+};
+
+namespace LittleEndian {
+
+class BufferWriter : public BufferWriterBase<BufferWriter>
+{
+public:
+    BufferWriter(uint8_t * buf, size_t len) : BufferWriterBase<BufferWriter>(buf, len) {}
+    BufferWriter(const BufferWriter & other) = default;
+    BufferWriter & operator=(const BufferWriter & other) = default;
+
+    BufferWriter & EndianPut(uint64_t x, size_t size)
+    {
+        while (size-- > 0)
+        {
+            uint8_t c = x & 0xff;
+            Put(c);
+            x >>= 8;
+        }
+        return *this;
+    }
+
+    /// Allows treating of a buffer stream to a different endianess.
+    /// If input buffer does not fit, return value will not fit either.
+    template <typename T>
+    static BufferWriter StartingFrom(BufferWriterBase<T> & other)
+    {
+        if (!other.Fit())
+        {
+            return BufferWriter(other.Buffer() + other.Size(), 0, other.Needed() - other.Size());
+        }
+        return BufferWriter(other.Buffer() + other.Needed(), other.Size() - other.Needed());
+    }
+
+private:
+    BufferWriter(uint8_t * buf, size_t len, size_t needed) : BufferWriterBase<BufferWriter>(buf, len, needed) {}
+};
+
+} // namespace LittleEndian
+
+namespace BigEndian {
+
+class BufferWriter : public BufferWriterBase<BufferWriter>
+{
+public:
+    BufferWriter(uint8_t * buf, size_t len) : BufferWriterBase<BufferWriter>(buf, len) {}
+    BufferWriter(const BufferWriter & other) = default;
+    BufferWriter & operator=(const BufferWriter & other) = default;
+
+    BufferWriter & EndianPut(uint64_t x, size_t size)
+    {
+        while (size-- > 0)
+        {
+            uint8_t c = (x >> (size * 8)) & 0xff;
+            Put(c);
+        }
+        return *this;
+    }
+
+    /// Allows treating of a buffer stream to a different endianess.
+    /// If input buffer does not fit, return value will not fit either.
+    template <typename T>
+    static BufferWriter StartingFrom(BufferWriterBase<T> & other)
+    {
+        if (!other.Fit())
+        {
+            return BufferWriter(other.Buffer() + other.Size(), 0, other.Needed() - other.Size());
```

Comment:
There seems to be a practical problem here:  If I write into a writer for a bit, then create a new one for the new endianness and write into that for a bit, how do I measure the total amount of buffer needed when I'm done?  Adding up the `Needed()` values from the two won't do the right thing if the first one doesn't fit (because we will get "# of bytes written to first one" + "# of bytes written to second one" + "# of bytes that did not fit in the first one) but will if the first one fit, right?

It would be good to get this part figured out.

Reply:
I am not sure how to evolve this one. 

Generally I find that mixing endianess is strange. Michael suggested a single base class that has both LE and BE calls.
I would suggest to iterate once we see usage. My guess is that we will never have mixed endianess.

I believe current logic is: `sum up needed from all, as long as last one Fit()`

Reply:
> I believe current logic is: sum up needed from all, as long as last one Fit()

But that fails to address the point of continuing even if stuff is not fitting, which is to figure out how much buffer you _would_ need to encode all your data.  That is, the "measure needed size" function is broken here.

> My guess is that we will never have mixed endianess.

I agree, and would somewhat prefer we not add code that is meant to handle that might-never-arise case, especially if that code has known problems...